vecindex: fix and enhance vecindex testing

andy-kimball opened the pull request:
#### vecindex: add datadriven tests for vecindex

Refactor the C-SPANN datadriven tests so they can be shared by other
Store implementations. Add new datadriven test for vecindex that
tests recall for several interesting datasets.

#### cspann: fix bug in CalculateTruth

Fix bug in CalculateTruth, which was calling vecpb.MeasureDistance with
un-normalized vectors for Cosine distance. This returns incorrect results.

#### vecindex: fix determinism violations

Fix a couple cases where the vector index is not behaving
deterministically even when configured to do so. In one case, partition
keys were not generated deterministically. In the other case, the seed
used in crdb_test builds was different than the seed used otherwise.

#### builtins: add process_vector_index_fixups builtin

Add a new builtin function that waits until all pending fixups have
been processed for a given index. This is used for testing.

Previously, we were processing fixups in deterministic tests by calling
ProcessFixups just before inserting vectors or searching for them.
However, that was causing deadlocks, because fixups were being processed
in the same transaction that was updating the index.

The vector_index test is now updated to call the new builtin when it
needs to force fixup processing.

#### cspann: use brackets to format vectors

Postgres uses brackets to format vectors, e.g. [1, 2]. However, our test
code currently uses parentheses, e.g. (1, 2). Update our tests to use
brackets instead, so that we're more consistent.

andy-kimball committed:
cspann: use brackets to format vectors

Postgres uses brackets to format vectors, e.g. [1, 2]. However, our test
code currently uses parentheses, e.g. (1, 2). Update our tests to use
brackets instead, so that we're more consistent.

Epic: CRDB-42943
Release note: None

andy-kimball committed:
builtins: add process_vector_index_fixups builtin

Add a new builtin function that waits until all pending fixups have
been processed for a given index. This is used for testing.

Previously, we were processing fixups in deterministic tests by calling
ProcessFixups just before inserting vectors or searching for them.
However, that was causing deadlocks, because fixups were being processed
in the same transaction that was updating the index.

The vector_index test is now updated to call the new builtin when it
needs to force fixup processing.

Epic: CRDB-42943
Release note: None

andy-kimball committed:
vecindex: fix determinism violations

Fix a couple cases where the vector index is not behaving
deterministically even when configured to do so. In one case, partition
keys were not generated deterministically. In the other case, the seed
used in crdb_test builds was different than the seed used otherwise.

Epic: CRDB-42943
Release note: None

andy-kimball committed:
vecindex: add datadriven tests for vecindex

Refactor the C-SPANN datadriven tests so they can be shared by other
Store implementations. Add new datadriven test for vecindex that
tests recall for several interesting datasets.

Epic: CRDB-42943
Release note: None

andy-kimball committed:
cspann: fix bug in CalculateTruth

Fix bug in CalculateTruth, which was calling vecpb.MeasureDistance with
un-normalized vectors for Cosine distance. This returns incorrect results.

Epic: CRDB-42943
Release note: None

mw5h reviewed (commented):
Looks good, thanks for doing this!

DrewKimball commented on pkg/sql/planner.go:
```diff
@@ -1076,3 +1076,15 @@ func (p *planner) ExtendHistoryRetention(ctx context.Context, jobID jobspb.JobID
 }
+
+// ProcessVectorIndexFixups is part of the eval.Planner interface.
+func (p *planner) ProcessVectorIndexFixups(
```
Off the top of my head, I'm not sure if there's something better. The whole planner / eval.Context thing is probably due for a refactor.
<!-- Reviewable comment -OWH3yDLDwdqxFqO-99q:b-d89xlq -->
<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 17 of 17 files at r1, 8 of 8 files at r2, 2 of 2 files at r3, 2 of 2 files at r4, 15 of 15 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150573#-:-OWHqgpr3TOiCHL-X-Gz:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mw5h)

___
*[`-- commits` line 20 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150573#-OWH3mhF28L3J_hE2h-g:-OWH3mhF28L3J_hE2h-h:b-l30adq):*
How were fixups happening in the same txn? Do you just mean they were interleaved?
___
*[`pkg/sql/vecindex/vecindex_test.go` line 185 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/150573#-OWHqDd09_2Czd9CX7PV:-OWHqDd09_2Czd9CX7PW:b-g9cq3u) ([raw file](https://github.com/cockroachdb/cockroach/blob/13141283589ad9f3f703520d9b71bf44021ad252/pkg/sql/vecindex/vecindex_test.go#L185)):*
> ```Go
> 	ti.runner.Exec(ti.T, fmt.Sprintf("SET vector_search_beam_size = %d", beamSize))
> 
> 	query := fmt.Sprintf(`SELECT id FROM t ORDER BY v %s $1 LIMIT %d`, op, topK*3)
> ```

Should this be multiplied by rerankMultiplier instead of 3?
___
*[`pkg/sql/planner.go` line 1088 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150573#-OWHmI_C6bQEldDYsCCl:-OWHmI_C6bQEldDYsCCm:brc0mby) ([raw file](https://github.com/cockroachdb/cockroach/blob/83c7b0e42a3a043057044bc6361e10411d68414a/pkg/sql/planner.go#L1088)):*
> ```Go
> 		return err
> 	}
> 	vi.ProcessFixups()
> ```

It's probably a good idea to make `ProcessFixups` check for context cancellation.
___
*[`pkg/sql/logictest/testdata/logic_test/vector_index` line 555 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150573#-OWHmXTD71E813N4UAxR:-OWHmXTEE_aQJzLxBXkd:b-8mdsbo) ([raw file](https://github.com/cockroachdb/cockroach/blob/83c7b0e42a3a043057044bc6361e10411d68414a/pkg/sql/logictest/testdata/logic_test/vector_index#L555)):*
> ```
> SELECT crdb_internal.process_vector_index_fixups(
>   (SELECT id FROM system.namespace WHERE name = 'exec_test'),
>   (SELECT index_id FROM crdb_internal.table_indexes WHERE descriptor_name = 'exec_test' AND index_name = 'idx1'));
> ```

nit: let's define a UDF that takes the table and index names and then calls the builtin.


<!-- Sent from Reviewable.io -->

andy-kimball commented:
*[`pkg/sql/planner.go` line 1088 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150573#-OWHmI_C6bQEldDYsCCl:-OWN-eKBDqBQETRgpBQq:b-apag5p) ([raw file](https://github.com/cockroachdb/cockroach/blob/83c7b0e42a3a043057044bc6361e10411d68414a/pkg/sql/planner.go#L1088)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

It's probably a good idea to make `ProcessFixups` check for context cancellation.
</blockquote></details>

The problem is that `context.Context` and `sync.Cond` don't really mix. The whole thing would probably need to be rewritten.


<!-- Sent from Reviewable.io -->

andy-kimball committed:
cspann: support context cancelation in Process

Previously, the Process method did not take a context parameter and so
couldn't be canceled. This was not a problem when it was only used in
testing, but now it can be called as a built-in function. This commit
adds a context parameter to the method and supports cancelation.

Epic: CRDB-42943
Release note: None

andy-kimball commented:
*[`pkg/sql/planner.go` line 1088 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150573#-OWHmI_C6bQEldDYsCCl:-OWoXj_uEmzEZ2WyEeof:bazv5uv) ([raw file](https://github.com/cockroachdb/cockroach/blob/83c7b0e42a3a043057044bc6361e10411d68414a/pkg/sql/planner.go#L1088)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

The problem is that `context.Context` and `sync.Cond` don't really mix. The whole thing would probably need to be rewritten.
</blockquote></details>

I ended up doing this work. I added another commit, PTAL.


<!-- Sent from Reviewable.io -->

mw5h commented on pkg/sql/vecindex/vecstore/store.go:
```diff
@@ -196,6 +208,11 @@ func (s *Store) RunTransaction(ctx context.Context, fn func(txn cspann.Txn) erro
 func (s *Store) MakePartitionKey() cspann.PartitionKey {
+	if s.isDeterministic {
+		// Generate new partition keys by incrementing an atomic.
+		return cspann.PartitionKey(s.keyGen.Add(1))
```
Will this work with a multi-node logic test?

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150573)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @mw5h)



<!-- Sent from Reviewable.io -->

andy-kimball commented:
Failure in extended CI is a [known issue](https://github.com/cockroachdb/cockroach/issues/146749).

andy-kimball commented:
bors r=drewkimball
